### DISCUSSION TITLE
Refactor airodump-ng.c

### DISCUSSION THREAD

**Discussion by gemesa** [aircrack-ng/aircrack-ng#2472] [Ideas]
I have been working on the codebase of `airodump-ng` lately and I think some functions are way too long and hard to maintain. So I am planning to do some refactoring if that is okay with you. I am interested also in your suggestions.

Some examples I thought of:
- replace magic numbers with defines
```
/* skip all non probe response frames in active scanning simulation mode */
	if (lopt.active_scan_sim > 0 && h80211[0] != 0x50) return (0);
```
-->
```
/* skip all non probe response frames in active scanning simulation mode */
	if (lopt.active_scan_sim > 0 && h80211[0] != IEEE80211_FC0_SUBTYPE_PROBE_RESP) return (0);
```
---
- split huge functions into smaller ones

```
static int dump_add_packet(unsigned char * h80211,
						   int caplen,
						   struct rx_info * ri,
						   int cardnum)
{
...
	if (getMACcount(lopt.rBSSID) > 0)
	{
		if (memcmp(opt.f_netmask, NULL_MAC, 6) != 0)
		{
			if (is_filtered_netmask(bssid)) return (1);
		}
		else
		{
			while (cur->next != NULL)
			{
				cur = cur->next;
				if (memcmp(cur->mac, bssid, 6) == 0)
				{
					match = 1;
					break;
				}
			}
			if (match != 1) return (1);
		}
	}
...
```
-->
```
uint8_t is_filtered_bssid(uint8_t* bssid)
{
	uint8_t match = 0;
	pMAC_t cur = lopt.rBSSID;
	
	if (memcmp(opt.f_netmask, NULL_MAC, 6) != 0)
	{
		if (is_filtered_netmask(bssid)) return (1);
	}
	else
	{
		while (cur->next != NULL)
		{
			cur = cur->next;
			if (memcmp(cur->mac, bssid, 6) == 0)
			{
				match = 1;
				break;
			}
		}
		if (match != 1) return (1);
	}

	return match;
}

static int dump_add_packet(unsigned char * h80211,
						   int caplen,
						   struct rx_info * ri,
						   int cardnum)
{
...
	if (getMACcount(lopt.rBSSID) > 0)
	{
		if (is_filtered_bssid(bssid)) return (1);
	}
...
```
---
- replace calls with aliases that are easy to understand
```
if (memcmp(ap_cur->bssid, BROADCAST, 6) != 0) {...}
```
-->
```
uint8_t is_associated_station(uint8_t* bssid)
{
	return memcmp(bssid, BROADCAST, 6) != 0;
}

if (is_associated_station(ap_cur->bssid)) {...}
```

**Comment by Mister-X-** [comment]
Sounds great.

Note that my bandwidth is limited at this time, which is the reason why some of your other PR haven't been merged yet (and that I haven't made a release in so long).

As for suggestions:
- A smaller PR is often always easier to merge that a large one, as it requires less time to review.
- If there is a specific order I should merge them, or if a PR depends on another, please indicate that in the PR.
- Documenting the functions, and if you can, the code.
- Possibly organize them in separate files (like it was done for output files)

Regarding the second function example, I feel like a `#define` would be a better fit in that case, as a function would take more CPU time than it's worth (look into `static` and `inline` if you want to use functions).

**Comment by Mister-X-** [comment]
I would have put it in #2495, but that's unrelated: Another thing that should be done prior to doing a lot of refactor, is adding automated testing to ensure the changes don't break the UI.

**Reply by gemesa** [reply]
> I would have put it in #2495, but that's unrelated: Another thing that should be done prior to doing a lot of refactor, is adding automated testing to ensure the changes don't break the UI.

Do you have any suggestions how to test the UI automatically?

**Reply by Mister-X-** [reply]
Yes. Running airodump-ng with different PCAP, and grabbing the output, then checking it against what we expect.

Using `timeout`, for example `sudo timeout 5 airodump-ng TEST.pcap`.

@jbenden Do you know any tool that we could integrate in the CI/CD for TUI testing?

**Comment by gemesa** [comment]
You are right. I was playing with resizing my terminal and testing `airodump-ng` like that and #2495 is not really working as I would expect, so I will close it for now.

**Comment by CorySpindolaWiH** [comment]
Great that you are planning to refactor the codebase of airodump-ng. Refactoring can improve the maintainability of the codebase and make it easier to add new features and fix bugs.

Your suggested refactoring ideas look good. Here are some additional suggestions:

Use meaningful variable names: Instead of using single letter variable names or generic names like match, use descriptive names that convey the meaning of the variable.

Remove dead code: Look for code that is not being used anymore and remove it. This can help simplify the code and make it easier to understand.

Add comments: Add comments to the code to explain what each function and block of code does. This can help others understand the code and make it easier to maintain.

Use enums: Instead of using defines, consider using enums to define constants. This can make the code more readable and easier to understand.

Use const: Use the const keyword to define variables that should not be modified. This can help prevent bugs and make the code more reliable.

Write unit tests: Write unit tests for the functions you refactor. This can help ensure that the refactoring does not introduce new bugs and that the code still works as expected.

Overall, the goal of refactoring is to make the code more readable, maintainable, and reliable. Good luck with your refactoring efforts!
